miden-crypto/src/merkle/smt/large_forest/history/mod.rs

```diff
@@ -0,0 +1,430 @@
+/// stored in the leaves of an otherwise shallower merkle tree.
+pub type CompactLeaf = Map<Word, Word>;
+
+/// A collection of changes to arbitrary non-leaf nodes in a merkle tree.
+///
+/// All changes to nodes between versions `v` and `v + 1` must be explicitly "undone" in the
+/// `NodeChanges` representing version `v`. This includes nodes that were defaulted in version `v`
+/// that were given an explicit value in version `v + 1`, where the `NodeChanges` must explicitly
+/// set those nodes back to the default.
+///
+/// Failure to do so will result in incorrect values when those nodes are queried at a point in the
+/// history corresponding to version `v`.
+pub type NodeChanges = Map<NodeIndex, Word>;
+
+/// A collection of changes to arbitrary leaf nodes in a merkle tree.
+///
+/// This represents the state of the leaf wholesale, rather than as a delta from the newer version.
+/// This massively simplifies querying leaves in the history.
+///
+/// Note that if in the version of the tree represented by these `LeafChanges` had the default value
+/// at the leaf, this default value must be made concrete in the map. Failure to do so will retain a
+/// newer, non-default value for that leaf, and thus result in incorrect query results at this point
+/// in the history.
+pub type LeafChanges = Map<LeafIndex<SMT_DEPTH>, CompactLeaf>;
+
+/// An identifier for a historical tree version overlay, which must be monotonic as new versions are
+/// added.
+pub type VersionId = u64;
+
+// HISTORY
+// ================================================================================================
+
+/// A History contains a sequence of versions atop a given tree.
+///
+/// The versions are _cumulative_, meaning that querying the history must account for changes from
+/// the current tree that take place in versions that are not the queried version or the current
+/// tree.
+#[derive(Clone, Debug)]
+pub struct History {
+    /// The maximum number of historical versions to be stored.
+    max_count: usize,
+
+    /// The deltas that make up the history for this tree.
+    ///
+    /// It will never contain more than `max_count` deltas, and is ordered with the oldest data at
+    /// the lowest index.
+    ///
+    /// # Implementation Note
+    ///
+    /// As we are targeting small numbers of history items (e.g. 30), having a sequence with an
+    /// allocated capacity equal to the small maximum number of items is perfectly sane. This will
+    /// avoid costly reallocations in the fast path.
+    ///
+    /// We use a [`VecDeque`] instead of a [`Vec`] or [`alloc::collections::LinkedList`] as we
+    /// estimate that the vast majority of removals will be the oldest entries as new ones are
+    /// pushed. This means that we can optimize for those removals along with indexing performance,
+    /// rather than optimizing for more rare removals from the middle of the sequence.
```

Comment:
I don't quite understand how we would be able to do so, unless your interpretation is always one-delta-to-any-step, so the `Delta` is an accumulation of a set of reversions derived from i.e. a _forward_ `MutationSet`

Reply:
Correct!